File: 0046-permutations/0046-permutations.py
class Solution:
    def fun(self, ind, nums, ans):
        if ind == len(nums):
            ans.append(nums[:])
            return

        for i in range(ind, len(nums)):
            self.swap(ind, i, nums)
            self.fun(ind+1, nums, ans)
            self.swap(ind, i, nums)

    def swap(self, i, j, nums):
        t = nums[i]
        nums[i] = nums[j]
        nums[j] = t

    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.fun(0, nums, ans)
        return ans

File: 0046-permutations/0046-permutations.java
class Solution {
    void fun(int a[], int start, List<List<Integer>> ans) {
        if (start == a.length) {
            List<Integer> permutation = new ArrayList<>();
            for (int num : a) {
                permutation.add(num);
            }
            ans.add(permutation);
            return;
        }
        for (int i = start; i < a.length; i++) {
            swap(a, start, i);
            fun(a, start + 1, ans);
            swap(a, start, i);
        }
    }

    void swap(int[] a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        fun(nums, 0, ans);
        return ans;
    }
}
